src/engraving/dom/rest.cpp

```diff
@@ -553,12 +611,40 @@ bool Rest::isWholeRest() const
 {
     TDuration durType = durationType();
     return durType == DurationType::V_WHOLE
-           || (durType == DurationType::V_MEASURE && measure() && measure()->ticks() < Fraction(2, 1));
+           || (isFullMeasureRest() && measure() && measure()->ticks() < Fraction(2, 1));
 }
 
-int Rest::computeNaturalLine(int lines) const
+bool Rest::isBreveRest() const
+{
+    TDuration durType = durationType();
+    return durType == DurationType::V_BREVE
+           || (isFullMeasureRest() && measure() && measure()->ticks() >= Fraction(2, 1));
+}
+
+int Rest::computeNaturalLine(DurationType type, int lines) const
 {
     int line = (lines % 2) ? floor(double(lines) / 2) : ceil(double(lines) / 2);
+
+    switch (type) {
+    case DurationType::V_WHOLE:
+        if (lines > 1) {
+            line--;
+        }
+        break;
+    case DurationType::V_BREVE:
+        if (lines <= 1) {
+            line++;
+        }
+        break;
+    case DurationType::V_MEASURE:
+        if (isBreveRest()) {
+            return computeNaturalLine(DurationType::V_BREVE, lines);
+        }
+        return computeNaturalLine(DurationType::V_WHOLE, lines);
```

Comment:
>This (recursion) is a big no-no

What? Recursion is a powerful language feature. Why would we eschew it? I can see that you and I have a fundamental philosophical disagreement here, so I will shut up and agree to disagree. But I feel pretty strongly that it is extremely useful with risks that can be mitigated.

I think a much bigger no-no is code duplication. Even one line. To me, the most important feature of a codebase is that each business rule be implemented exactly once. The MuseScore codebase is quite good in this regard, but it could be even better. The `V_MEASURE` duration type is an example. Graphically it is either a whole or a breve. I hate to see code duplicated to calculate one or the other.

Furthermore, I honestly think `switch` fall-thrus are a worse no-no than recursion.

Reply:
Recursion is absolutely not necessarily bad in general; and in this case also not necessarily, but it's perhaps not ideal yet. For one thing, because `int line` is calculated twice; the first time, it is discarded because of the recursive call.
It might be even better to extract the part that determines which symbol is used, so that it becomes something like this:
```c++
    if (isBreveRest()) {
        if (lines <= 1) {
             line++;
         }
    } else if (isWholeRest()) {
        if (lines > 1) {
             line--;
         }
    }
```

We could even consider creating a separate enum for rest symbols (or using SymIds), and converting from DurationType to that enum, and then doing a switch (without recursion) over the cases of that enum. That might be too much though.

Reply:
Another option would be a lamda function. I switched from recursion to a lamda in another PR specifically because of this commetn.